src/Workspaces/Remote/Core/Diagnostics/DiagnosticComputer.cs

```diff
@@ -124,28 +177,102 @@ private static ImmutableArray<DiagnosticAnalyzer> GetAnalyzers(BidirectionalMap<
             return builder.ToImmutable();
         }
 
-        private static BidirectionalMap<string, DiagnosticAnalyzer> CreateAnalyzerMap(Project project)
+        private static async Task<(CompilationWithAnalyzers compilationWithAnalyzers, BidirectionalMap<string, DiagnosticAnalyzer> analyzerToIdMap)> GetOrCreateCompilationWithAnalyzersAsync(
+            Project project,
+            bool isDocumentAnalysis,
+            CancellationToken cancellationToken)
+        {
+            var cacheEntry = await GetOrCreateCacheEntryAsync().ConfigureAwait(false);
+            return (cacheEntry.CompilationWithAnalyzers, cacheEntry.AnalyzerToIdMap);
+
+            async Task<CompilationWithAnalyzersCacheEntry> GetOrCreateCacheEntryAsync()
+            {
+                if (!isDocumentAnalysis)
+                {
+                    // Only use cache for document analysis.
+                    return await CreateCompilationWithAnalyzersCacheEntryAsync(project, cancellationToken).ConfigureAwait(false);
+                }
```

Comment:
One more tweak possible: if we move this check to GetDiagnosticsAsync and change CreateCompilationWithAnalyzersCacheEntryAsync to return the tuple, we'll end up with standard GetOrCreate pattern for CWT.

Reply:
That was making the code more muddled, I'll stick with the current form.